apps/cms.c

```diff
@@ -885,10 +892,29 @@ int cms_main(int argc, char **argv)
             goto end;
     }
 
-    in = bio_open_default(infile, 'r',
-                          binary_files ? FORMAT_BINARY : informat);
-    if (in == NULL)
-        goto end;
+    if (operation != SMIME_SIGN && digesthex != NULL) {
+        BIO_printf(bio_err,
+                   "Warning: -digest is ignored for non-signing operation\n");
+        digesthex = NULL;
+    }
+    if (digesthex != NULL) {
+        if (infile != NULL || !(flags & CMS_DETACHED)) {
```

Comment:
There is one more option that causes the precomputed digest to not be used and that is the CMS_STREAM. But please note what happens at lines 1071-1073 - for the SMIME output format the CMS_STREAM is automatically enabled meaning the precomputed digest won't be used! That looks to me like a problem.

Reply:
I can add CMS_STREAM to the error check on lines 901-903.

Can we skip automatically enabling CMS_STREAM on line 1073 if -digest is given?

Reply:
> I can add CMS_STREAM to the error check on lines 901-903.

Please do.

> Can we skip automatically enabling CMS_STREAM on line 1073 if -digest is given?

Please try that and check whether the produced output makes any sense. If so, then please do that and add testcases.

Reply:
It works!